packages/@lwc/integration-types/src/decorators/wire.ts

```diff
@@ -0,0 +1,404 @@
+    // @ts-expect-error Too many wire parameters
+    @wire(TestAdapter, { config: 'config' }, {})
+    tooManyWireParams?: TestValue;
+    // @ts-expect-error Bad config type
+    @wire(TestAdapter, { bad: 'value' })
+    badConfig?: TestValue;
+    // @ts-expect-error Bad prop type
+    @wire(TestAdapter, { config: 'config' })
+    badPropType?: { bad: 'value' };
+    // @ts-expect-error Prop must be optional or assigned in constructor
+    @wire(TestAdapter, { config: 'config' }) notOptional: TestValue;
+    // @ts-expect-error Referenced reactive prop does not exist
+    @wire(TestAdapter, { config: '$nonexistentProp' } as const)
+    nonExistentReactiveProp?: TestValue;
+    // @ts-expect-error Referenced reactive prop is the wrong type
+    @wire(TestAdapter, { config: '$number' } as const)
+    numberReactiveProp?: TestValue;
+    // @ts-expect-error Referenced nested reactive prop does not exist
+    @wire(TestAdapter, { config: '$nested.nonexistent' } as const)
+    nonexistentNestedReactiveProp?: TestValue;
+    // @ts-expect-error Referenced nested reactive prop does not exist
+    @wire(TestAdapter, { config: '$nested.invalid' } as const)
+    invalidNestedReactiveProp?: TestValue;
+    // @ts-expect-error Incorrect non-reactive string literal type
+    @wire(TestAdapter, { config: 'not reactive' } as const)
+    nonReactiveStringLiteral?: TestValue;
+    // @ts-expect-error Nested props are not reactive - only top level
+    @wire(DeepConfigAdapter, { deep: { config: '$number' } } as const)
+    deepReactive?: TestValue;
+    // @ts-expect-error Looks like a method, but it's actually a prop
+    @wire(TestAdapter, { config: 'config' })
+    propValueIsMethod = function (this: PropertyDecorators, _: TestValue): void {};
+
+    // --- AMBIGUOUS --- //
+    // Passing a config is optional because adapters don't strictly need to use it.
+    // Can we be smarter about the type and require a config, but only if the adapter does?
+    @wire(TestAdapter)
+    noConfig?: TestValue;
+    // Because the basic type `string` could be _any_ string, we can't narrow it and compare against
+    // the component's props, so we must accept all string props, even if they're incorrect.
+    // We could technically be strict, and enforce that all configs objects use `as const`, but very
+    // few projects currently use it (there is no need) and the error reported is not simple to
+    // understand.
+    @wire(TestAdapter, { config: 'incorrect' })
+    wrongConfigButInferredAsString?: TestValue;
+    // People shouldn't do this, and they probably never (heh) will. TypeScript allows it, though.
+    @wire(TestAdapter, { config: 'config' })
+    never?: never;
+}
+
+/** Validations for decorated methods */
+export class MethodDecorators extends LightningElement {
+    // Helper props
+    configProp = 'config' as const;
+    nested = { prop: 'config', invalid: 123 } as const;
+    'nested.prop' = false; // should be unused
```

Comment:
This is kind of ambiguous because it's not clear whether `$nested.prop` refers to line 130 or 131. Better to rename this?

Reply:
This is deliberate, to validate that `$nested.prop` picks up the object, rather than the weird prop name. I'll update the comment to clarify